bootloaders/uart/uart_boot.c

```diff
@@ -0,0 +1,218 @@
+            printf("Write has hung - resetting\n");
+            reset_chip();
+            return;
+        }
+        in = uart_getc(UART_ID);
+        if (in != 'w') {
+            printf("Incorrect response - resetting\n");
+            reset_chip();
+            return;
+        }
+        current_addr += 32;
+    }
+
+    uint32_t time_end = time_us_32();
+    printf("Write took %dus\n", time_end - time_start);
+    printf("Write complete - resetting pointer\n");
+
+    uart_putc_raw(UART_ID, 'c');
+    if (!uart_is_readable_within_us(UART_ID, 500)) {
+        // Detect hangs and reset the chip
+        printf("Clear has hung - resetting\n");
+        reset_chip();
+        return;
+    }
+    in = uart_getc(UART_ID);
+    printf("%c\n", in);
+    if (in != 'c') {
+        printf("Incorrect response - resetting\n");
+        reset_chip();
+        return;
+    }
+
+    printf("Verifying binary\n");
+    time_start = time_us_32();
+    current_addr = start_addr;
+    while (current_addr < end_addr) {
+        uart_putc_raw(UART_ID, 'r');
+        char *buf = (char*)current_addr;
+        if (!uart_is_readable_within_us(UART_ID, 500)) {
+            // Detect hangs and reset the chip
+            printf("Verify has hung - resetting\n");
+            reset_chip();
+            return;
+        }
+        int i = 0;
+        while (uart_is_readable_within_us(UART_ID, 10) && i < 32) {
+            in = uart_getc(UART_ID);
+            if (in != buf[i]) {
+                printf("Verify has incorrect data at 0x%08x - resetting\n", current_addr - start_addr + SRAM_BASE);
+            }
+            i++;
+        }
+        if (i != 32) {
+            printf("Verify has incorrect data size - resetting\n");
+        }
+        in = uart_getc(UART_ID);
+        if (in != 'r') {
+            printf("Incorrect response - resetting\n");
+            reset_chip();
+            return;
+        }
+        current_addr += 32;
+    }
+
+    time_end = time_us_32();
+    printf("Verify took %dus\n", time_end - time_start);
+    printf("Verify complete - executing\n");
+
+    uart_putc_raw(UART_ID, 'x');
+    if (!uart_is_readable_within_us(UART_ID, 500)) {
+        // Detect hangs and reset the chip
+        printf("Execute has hung - resetting\n");
+        reset_chip();
+        return;
+    }
+    in = uart_getc(UART_ID);
+    printf("%c\n", in);
+    if (in != 'x') {
+        printf("Incorrect response - resetting\n");
+        reset_chip();
+        return;
+    }
+}
+
+
+int main()
+{
+    stdio_init_all();
+
+    // Set up our UART for booting the other device
+    uart_init(UART_ID, 1000000);
+    gpio_set_function(UART_TX_PIN, GPIO_FUNC_UART);
+    gpio_set_function(UART_RX_PIN, GPIO_FUNC_UART);
+
+    // Set up run pin
+    gpio_init(RUN_PIN);
+    gpio_set_dir(RUN_PIN, GPIO_OUT);
+
+    // Reset chip
+    reset_chip();
+
+
+    while (true) {
+        char splash[] = "RP2350";
+        char hello[] = "Hello";
```

Comment:
minor nitpick: I guess these two variables could be moved outside of the `while (true)` loop?